security/audit: More audit enhancements

graphcareful opened the pull request:
This PR compiles a few unrelated auditing changes into a single PR.  Listed, they are:
- Integration with `v/feature` so auditing can only truly be enabled when the cluster is fully upgraded: [#868](https://github.com/redpanda-data/core-internal/issues/868)
- Best effort attempt to drain audit log on clean shutdown of node or disable of audit feature flag: [#901](https://github.com/redpanda-data/core-internal/issues/901)
- Bugfix preventing copies of OCSF structures
- Implemented audit queue backpressure by size bytes removing previous method of accounting by raw number of elements enqueued: [#14099](https://github.com/redpanda-data/redpanda/issues/14099)
- Misc. test refactor 

Fixes: #14099 
Fixes: https://github.com/redpanda-data/core-internal/issues/868
Fixes: https://github.com/redpanda-data/core-internal/issues/901

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

### Bug Fixes

* Fixed a bug where OCSF messages were being copied instead of moved

### Improvements

* Changes `audit_max_queue_elements_per_shard` config option to `audit_queue_max_buffer_size_per_shard` for accounting how much memory to allocate per shard for queued accounting messages
* Auditing will now try to drain messages as best it can before clean shutdown or disable of the `audit_enabled` feature flag

dotnwat commented on src/v/security/audit/schemas/schemas.h:
```diff
@@ -176,5 +182,49 @@ class ocsf_base_event : public ocsf_base_impl {
+                sz += estimated_ocsf_size(element);
+            }
+        } else if constexpr (reflection::is_std_optional<T>) {
+            sz += 1;
```
this is `sz += sizeof(std::optional<T>)` or something like that.

dotnwat commented on src/v/security/audit/schemas/schemas.h:
```diff
@@ -176,5 +182,49 @@ class ocsf_base_event : public ocsf_base_impl {
+                sz += estimated_ocsf_size(*t);
+            }
+        } else if constexpr (std::is_same_v<ss::sstring, T>) {
+            sz += t.size();
```
how precise do you want to be? because this will be off by 40 or so bytes if the string is long enough so you'd want `t.size() + sizeof(ss::sstring)`

dotnwat commented on src/v/security/audit/schemas/schemas.h:
```diff
@@ -176,5 +182,49 @@ class ocsf_base_event : public ocsf_base_impl {
+    /// not include in its return value for 'equality_fields()'. Currently
+    /// the only example of this is the 'network_endpoint' struct.
+    template<typename T>
+    size_t estimated_ocsf_size(const T& t) const noexcept {
```
why doesn't each ocsf type introduce a `size()` method?

graphcareful replied:
I would still need to iterate over all "fields" in an ocsf struct to get a total size. Also there are many ocsf types, all of them would need a size field.

michael-redpanda commented on src/v/features/feature_table.h:
```diff
@@ -350,6 +351,12 @@ constexpr static std::array feature_schema{
     "cloud_metadata_cluster_recovery",
     feature::cloud_metadata_cluster_recovery,
     feature_spec::available_policy::always,
+    feature_spec::prepare_policy::always},
+  feature_spec{
+    cluster::cluster_version{11},
+    "audit_logging",
+    feature::audit_logging,
+    feature_spec::available_policy::always,
```
Like GSSAPI (and I think some other features), I think we want to make sure we 'nag' the user if this is enabled and they don't have a license

graphcareful replied:
I'm pretty sure that is only for features which are "enterprise only" that are enabled without a license.

michael-redpanda replied:
Isn't auditing enterprise only?

graphcareful replied:
Confirmed this is the case, +1 will change

michael-redpanda commented on src/v/security/audit/audit_log_manager.cc:
```diff
@@ -390,18 +406,37 @@ ss::future<> audit_client::create_internal_topic() {
+    auto abort = ss::sleep_abortable(3s, exit)
+                   .then([this] {
+                       vlog(
+                         adtlog.info,
```
should this be a warn and indicate that it was due to timeout?

michael-redpanda commented on src/v/security/audit/audit_log_manager.cc:
```diff
@@ -390,18 +406,37 @@ ss::future<> audit_client::create_internal_topic() {
     }
     vlog(adtlog.info, "Shutting down audit client");
+    ss::abort_source exit;
+    auto abort = ss::sleep_abortable(3s, exit)
```
maybe I'm missing it, but if the queue drain interval is >3 seconds, is it possible that this timeout will complete before a drain interval is kicked off?  Does something force the drain interval to kick off?

graphcareful replied:
So the flow would be:
1. `audit_sink::stop()`
2. `audit_sink::do_toggle(false)`
3. `audit_log_manager::pause()`
4. Pause will drain all shard queues
5. When `pause()` returns `stop()` will be called on the `audit_client`

graphcareful replied:
If you have client buffering enabled too on call to `stop()` the client will attempt to flush its buffers as well , even if the thresholds haven't been met: 
https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/client/produce_partition.h#L56

graphcareful replied:
I think there may be an issue though @BenPope , this just waits 3s for this gate to close, but i think really the intention is that `_client.stop()` will cause the produce_partition to kick off a consume ignoring the buffering options (the line of code i linked above).

But with the changes I've made to the client I believe that once the client hits `send()` it will just fail because the newly added abort source was already triggered. Ideas?

I think the issue are the semantics. With the abort source it indicates "stop what your doing now" but the way the code was originally wrote, `stop()` behaved in a way that would "send whatever your have batched now"

michael-redpanda commented on src/v/security/audit/audit_log_manager.h:
```diff
@@ -205,6 +205,10 @@ class audit_log_manager
     /// audit topic.
     bool _auth_misconfigured{false};
 
+    /// Represents whether the feature is actually active, not the
+    /// representation of the config variable
+    bool _effectively_enabled{false};
+
```
does this factor into whether or not audit messages are enqueued or not?

graphcareful replied:
As it stands, no. I'm still on the fence about if it should or not

michael-redpanda replied:
I _feel_ like it should... what's the argument against it?  If it's "effectively" disabled, I'd say we don't enqueue and return true?

graphcareful replied:
well it could be effectively disabled but the audit switch is ON, and maybe the client thought "oh its enabled i assume from this point on messages are now being audited". I guess we could just assert in the docs its enabled when u see a relevant log line and it may take some time.

Maybe it is something we should address as proposed. The common case is the node starting up and then enabling auditing and the kafka client taking some time to complete the initialization phase, which sets up credentials and attempts to make the audit topic. If it cannot create the audit topic and its not already up, it will forever retry this, which means the switch will be in the ON state but effectively enabled is false for a long period of time.

michael-redpanda commented on src/v/security/audit/schemas/schemas.h:
```diff
@@ -37,6 +37,12 @@ type_uid get_ocsf_type(class_uid class_uid, T activity_id) {
+    ocsf_base_impl() = default;
+    ocsf_base_impl(ocsf_base_impl&&) = default;
+    ocsf_base_impl& operator=(ocsf_base_impl&&) = default;
+    ocsf_base_impl(const ocsf_base_impl&) = delete;
+    ocsf_base_impl& operator=(const ocsf_base_impl&) = delete;
```
👍

michael-redpanda commented on src/v/security/audit/schemas/schemas.h:
```diff
@@ -176,5 +182,49 @@ class ocsf_base_event : public ocsf_base_impl {
         return h;
     }
+
+    /// Method to estimate the size of an ocsf message
```
this is estimating the size of the serialized message, correct?  Does this take into account the size of the name of the object or just the value of the object

graphcareful replied:
No this is estimating the size of the object in memory, since the message is not yet serialized for the time it exists within the queue.

michael-redpanda replied:
Thanks for the clarification!

michael-redpanda commented on src/v/security/audit/schemas/tests/ocsf_schemas_test.cc:
```diff
@@ -1066,3 +1066,45 @@ BOOST_AUTO_TEST_CASE(make_authn_event_failure) {
+                      + sizeof(user.type_id);
+
+    BOOST_CHECK_EQUAL(estimated_size, 225);
+    BOOST_CHECK_EQUAL(authn.estimated_size(), 225);
```
can you add in a check here to see how close the estimated size is compared to the serialized JSON object?  Is a good tolerance +/- 10%? +/- 500B?

graphcareful replied:
Check comment above

michael-redpanda commented on src/v/config/configuration.cc:
```diff
@@ -282,7 +282,7 @@ configuration::configuration()
        .visibility = visibility::tunable},
       std::nullopt,
-      {.min = 32_MiB})
+      {.min = 1_MiB})
```
did you mean to change `raft_max_recovery_memory` min?  Should someone from replication verify this is ok?

graphcareful replied:
Nice catch this is definitely a git/texteditor error, meant to change an audit config

michael-redpanda replied:
This change is still present, did you mean to revert the change to `raft_max_recovery_memory`?

graphcareful replied:
Resolving now , latest force push addressed , nice catch

michael-redpanda reviewed (commented):
I think a good test for the 'best effort to drain' change is to see if you see the application lifecycle stop message from the audit system when it's stopped

dotnwat commented on src/v/security/audit/schemas/tests/ocsf_schemas_test.cc:
```diff
@@ -1066,3 +1066,45 @@ BOOST_AUTO_TEST_CASE(make_authn_event_failure) {
     BOOST_REQUIRE_EQUAL(ser, ::json::minify(expected));
 }
+
+BOOST_AUTO_TEST_CASE(test_ocsf_size) {
```
can we add a unit test for each type so that it triggers if someone adds something new to a structure in a way that is incompatible?

graphcareful replied:
So if someone adds a new OCSF struct to an OCSF message and it doesn't have an `equality_fields()` method, this will be a compilation error. Is that along the lines of what you were thinking?

dotnwat replied:
presumably

graphcareful replied:
Yeah thats also why i didn't make this a free floating method so you couldn't just pass any T to it.

michael-redpanda commented on src/v/security/audit/audit_log_manager.cc:
```diff
@@ -592,10 +589,11 @@ audit_log_manager::audit_log_manager(
       config::shard_local_cfg().audit_enabled_event_types.bind())
+  , _max_queue_size_bytes(
+      config::shard_local_cfg().audit_queue_max_buffer_size_per_shard())
+  , _queue_bytes_sem(_max_queue_size_bytes, "s/audit/buffer")
```
At first glance I thought you were putting in a regex

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/bc7cb15598aebe0d90bb9466f972d5b4c5bbb434..53e42e0b5e98c64af04f54fc3723664cf61bd188)
- Modified `audit_log_manager::stop` to also drain its queue one last time before shutdown
- Added a DT test to search for application activity stop message on node shutdown

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/53e42e0b5e98c64af04f54fc3723664cf61bd188..ea79ec3c650ce862562c87127eb1b868e3c8581a)
- Mark auditing as enterprise only

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/ea79ec3c650ce862562c87127eb1b868e3c8581a..aa8068bb2e011501c1ff47ce12ddffe1a6ecd546)
- Reverted change that accidentally modified a value for a raft config
- Changed default of `audit_client_max_buffer_size` to be a multiple of the new audit queue max size mb parameter

michael-redpanda reviewed (dismissed):
lgtm

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/41435#018bee36-0d58-4077-807d-b9eabab13b00:
```
"rptest.tests.audit_log_test.AuditLogTestsAppLifecycle.test_drain_on_shutdown"
```
<!--981d7179-61fb-441d-bc28-c152e6a34535-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/41435#018bee36-0d5b-4900-91d7-de959799bd68
<!--8857d8e5-59a8-4e2d-a78d-1be81983a25b-->

michael-redpanda reviewed (approved):
lgtm
